### PR TITLE
feat: add openziti support

### PR DESCRIPTION
**Pull request by dovholuknf**
Add OpenZiti support to EdgeX Foundry's compose project

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)

## Testing Instructions
* start OpenZiti: `docker compose -f openziti.yml --project-name edgex up`
* start EdgeX with modified docker compose `docker compose -f docker-compose.yml.worked up`
* access UI at http://localhost:4000

### CODE REVIEW

File: openziti-init.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FROM openziti/ziti-cli:0.33.1
+
+COPY ./openziti-init-entrypoint.sh /entrypoint.sh
+
+ENTRYPOINT [ "/entrypoint.sh" ]
+
```
Comment (lenny-intel) [inline openziti-init.Dockerfile:6]: We don't need a dockerfile to specify the entry point. This can be specified in the compose file.

File: openziti-init-entrypoint.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+makeEdgexService 'core-data'
+makeEdgexService 'core-metadata'
+makeEdgexService 'device-virtual'
+makeEdgexService 'device-rest'
+makeEdgexService 'rules-engine'
+makeEdgexService 'support-notifications'
+makeEdgexService 'support-scheduler'
+makeEdgexService 'sys-mgmt-agent'
+makeEdgexService 'ui'
+makeEdgexService 'app-rules-engine'
```
Comment (lenny-intel) [inline openziti-init-entrypoint.sh:54]: So OpenZiti has to know ahead of time the valid services?
Humm, how do we handle when user is running other edgex services. 
In this static case (not generated by compose builder) we need to specify call for all possible edgex services.
See all the `add-app*`, `add-asc*`, `add-device*` for complete list.
https://github.com/edgexfoundry/edgex-compose/tree/main/compose-builder

File: openziti-init-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ziti edge update identity edgex.rules-engine -a edgex.rules-engine-servers,edgex.core-command-clients
+ziti edge update identity edgex.device-virtual -a edgex.device-virtual-servers,edgex.core-command-clients,edgex.core-data-clients,edgex.core-metadata-clients,edgex.device-virtual-clients,edgex.rules-engine-clients,edgex.support-notifications-clients,edgex.support-scheduler-clients,edgex.sys-mgmt-agent-clients
+ziti edge update identity edgex.ui -a edgex.device-virtual-servers,edgex.core-command-clients,edgex.core-data-clients,edgex.core-metadata-clients,edgex.device-virtual-clients,edgex.rules-engine-clients,edgex.support-notifications-clients,edgex.support-scheduler-clients,edgex.sys-mgmt-agent-clients
+ziti edge update identity edgex.device-rest -a edgex.device-rest-servers,edgex.core-command-clients,edgex.core-data-clients,edgex.core-metadata-clients,edgex.device-virtual-clients,edgex.rules-engine-clients,edgex.support-notifications-clients,edgex.support-scheduler-clients,edgex.sys-mgmt-agent-clients
+ziti edge update identity edgex.core-command -a edgex,edgex.core-command-servers,edgex.core-metadata-clients,edgex.device-virtual-clients
```
Comment (lenny-intel) [inline openziti-init-entrypoint.sh:62]: Can these be part of makeEdgexService so we don't have to also repeat this for all possible edgex services.

File: openziti-init-entrypoint.sh [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+makeEdgexService 'device-rest'
+makeEdgexService 'rules-engine'
+makeEdgexService 'support-notifications'
+makeEdgexService 'support-scheduler'
+makeEdgexService 'sys-mgmt-agent'
+makeEdgexService 'ui'
+makeEdgexService 'app-rules-engine'
+
+
+ziti edge create identity edgexuser -P "${edgex_auth_policy}" --external-id edgexuser -a edgex.device-virtual-servers,edgex.core-command-clients,edgex.core-data-clients,edgex.core-metadata-clients,edgex.device-virtual-clients,edgex.rules-engine-clients,edgex.support-notifications-clients,edgex.support-scheduler-clients,edgex.sys-mgmt-agent-clients
```
Comment (lenny-intel) [inline openziti-init-entrypoint.sh:57]: edgex.sys-mgmt-agent-clients doesn't exist. Should be removed.

File: openziti-init-entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ziti edge update identity edgex.ui -a edgex.device-virtual-servers,edgex.core-command-clients,edgex.core-data-clients,edgex.core-metadata-clients,edgex.device-virtual-clients,edgex.rules-engine-clients,edgex.support-notifications-clients,edgex.support-scheduler-clients,edgex.sys-mgmt-agent-clients
+ziti edge update identity edgex.device-rest -a edgex.device-rest-servers,edgex.core-command-clients,edgex.core-data-clients,edgex.core-metadata-clients,edgex.device-virtual-clients,edgex.rules-engine-clients,edgex.support-notifications-clients,edgex.support-scheduler-clients,edgex.sys-mgmt-agent-clients
+ziti edge update identity edgex.core-command -a edgex,edgex.core-command-servers,edgex.core-metadata-clients,edgex.device-virtual-clients
+
+echo "granting additional dial capabilities to app-rules-engine"
+ziti edge update identity edgex.app-rules-engine -a edgex,edgex.app-rules-engine-servers,edgex.core-metadata-clients
```
Comment (lenny-intel) [inline openziti-init-entrypoint.sh:65]: Will this be needed for all app services?
Move this to `makeEdgexService`, add 2nd parameter for service type (`core` `support`, `app` device`) and only do this if `type=app`

File: openziti-init-entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+echo "auth policy id : ${edgex_auth_policy}"
+echo " "
+
+echo "creating the healthcheck proxy identity"
+ziti edge create identity \
+  healthcheck -o ${OPENZITI_PERSISTENCE_PATH}/healthcheck.jwt \
+  -a 'edgex.app-rules-engine-clients,edgex.device-rest-clients,edgex.core-command-clients,'\
+'edgex.core-data-clients,edgex.core-metadata-clients,edgex.device-virtual-clients,'\
+'edgex.rules-engine-clients,edgex.support-notifications-clients,edgex.support-scheduler-clients,'\
+'edgex.sys-mgmt-agent-clients'
```
Comment (lenny-intel) [inline openziti-init-entrypoint.sh:78]: Again we need to move this into `makeEdgexService` so we don't have to list out every service again.

File: openziti.yml [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    entrypoint:
+    - bash
+    - -euc
+    - |
+      ZITI_CMD+=" --ctrl-address ${EXTERNAL_DNS:-openziti}"\
+      " --ctrl-port ${ZITI_CTRL_EDGE_ADVERTISED_PORT:-1280}"\
+      " --router-address ${EXTERNAL_DNS:-openziti}"\
+      " --router-port ${ZITI_ROUTER_PORT:-3022}"\
+      " --password ${ZITI_PWD:-admin}"
+      exec ziti "$${@}" $${ZITI_CMD}
```
Comment (lenny-intel) [inline openziti.yml:22]: Put this at the bottom of your entry point script.

File: openziti.yml [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      edgex-network:
+        aliases:
+          - app-rules-engine.edgex.ziti
+          - core-command.edgex.ziti
+          - core-data.edgex.ziti
+          - core-metadata.edgex.ziti
+          - device-rest.edgex.ziti
+          - device-virtual.edgex.ziti
+          - support-notifications.edgex.ziti
+          - support-scheduler.edgex.ziti
```
Comment (lenny-intel) [inline openziti.yml:86]: Another place where the service need to all be listed. Need something more dynamic.

File: openziti.yml [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    image: openziti/ziti-cli:0.33.1
+    restart: unless-stopped
+    networks:
+      edgex-network:
+        aliases:
+          - openziti
+          - ziti-controller
+          - ziti-edge-controller
+          - ziti-router
+    entrypoint:
```
Comment (lenny-intel) [inline openziti.yml:13]:
```suggestion
    entrypoint:
    - /files/openziti-init-entrypoint.sh
```

File: openziti.yml [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    environment:
+      HOME: /edgex_openziti
+      OPENZITI_HEALTHCHECK_ALLOWED_PATH: '^.*/ping$'
+      OPENZITI_HEALTHCHECK_SEARCH_REGEX: '(.*).edgex.ziti'
+      OPENZITI_HEALTHCHECK_REPLACE_REGEX: 'edgex.$1'
+      OPENZITI_HEALTHCHECK_PROXY_PORT: 80
+      OPENZITI_HEALTHCHECK_IDENTITY: /edgex_openziti/healthcheck.json
+    volumes:
+      - edgex_openziti:/edgex_openziti
+    command: >
```
Comment (lenny-intel) [inline openziti.yml:68]: Create volume mount here to local file system in order to access the entry point script

File: openziti.yml [hunk 11]
```diff
@@ -0,0 +1,93 @@
+services:
```
Comment (lenny-intel) [inline openziti.yml:1]: rename this file to `docker-compose-openziti.yml`

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Makefile need the new `openziti` target similar to the `portainer` targets.

File: openziti-init.Dockerfile (hunk 1 above)
Comment (dovholuknf) [reply openziti-init.Dockerfile:6]: it exists to copy the script into the container but maybe you know of a better way to accomplish that? or i could bind mount the script, but figured this was as good?

File: openziti-init-entrypoint.sh (hunk 3 above)
Comment (dovholuknf) [reply openziti-init-entrypoint.sh:62]: since each one has individual, and discrete needs, I wanted to highligh only allowing communication from the services that are expected to communicate with other services. It's always easier to "let all the services talk to all the services", but I opted to follow the principle of least privilege here. 

I'll accomodate EdgeX however is best

File: openziti-init.Dockerfile (hunk 1 above)
Comment (lenny-intel) [reply openziti-init.Dockerfile:6]: Where is this Dockerfile being used?

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: openziti-init-entrypoint.sh (hunk 3 above)
Comment (lenny-intel) [reply openziti-init-entrypoint.sh:62]: Yes, I am ok if you want to do this for every edgex App and Device  service ...

File: openziti-init.Dockerfile (hunk 1 above)
Comment (lenny-intel) [reply openziti-init.Dockerfile:6]: OK, I see how you are using this. I like it.!

File: openziti.yml (hunk 9 above)
Comment (lenny-intel) [reply openziti.yml:13]: Nevermind

File: openziti.yml (hunk 10 above)
Comment (lenny-intel) [reply openziti.yml:68]: nevermind

File: openziti-init-entrypoint.sh (hunk 5 above)
Comment (dovholuknf) [reply openziti-init-entrypoint.sh:65]: it does all kind of "depend". Will all app-services receive a jwt from Vault they can use to authenticate to the overlay with? We can certainly make it easier by creating a "any edgex identity can communicate to any edgex identity" scheme if that's desired. This might end up needing to be a live conversation.

Comment (lenny-intel) [reply openziti-init-entrypoint.sh:65]: Yes, all App Service are just like App Rules Engine in this respect. Same with all Device services

File: openziti-init-entrypoint.sh (hunk 2 above)
Comment (dovholuknf) [reply openziti-init-entrypoint.sh:54]: OpenZiti has a whole API to use to provision things dynamically if needed. From what I saw in my travels, that wasn't needed but I don't have the whole scope of EdgeX in my head so this is how I did it. I can find the whole list and update accordingly

File: openziti-init-entrypoint.sh (hunk 3 above)
Comment (dovholuknf) [reply openziti-init-entrypoint.sh:62]: I can but, are there broad rules that are applicable to all "app-*" services? Like any app-service can talk to any core service? I need to know which services are authorized to talk to which services. I can't do that through trial and error.

File: openziti-init-entrypoint.sh (hunk 2 above)
Comment (lenny-intel) [reply openziti-init-entrypoint.sh:54]: Ya, for this first static pass. Then look at how to make it more dynamic with input from the user.

File: openziti-init-entrypoint.sh (hunk 3 above)
Comment (lenny-intel) [reply openziti-init-entrypoint.sh:62]: App Services can talk to any core/support service
Device Services will only talk to Core MetaData

File: openziti.yml (hunk 7 above)
Comment (lenny-intel) [reply openziti.yml:22]: nevermind

File: openziti-init-entrypoint.sh (hunk 5 above)
Comment (dovholuknf) [reply openziti-init-entrypoint.sh:65]: OpenZiti has attributes we can use. I can refactor these permsisions to use attributes like:
* core-services
* app-services
* devices-services

Any app-service or device-service would be authorized to communicate to any core-service if that's what you'd like to see. Should all app-services be authorized to commuicated to all app-services, all device-services? That is certainly easier mentally, but does reduce the security posture somewhat?

Or doe all services have the authorization to communicate to core-metadata? That's the only thing app-rules-engine is allowed to communicate to per this policy.

File: openziti.yml (hunk 7 above)
Comment (dovholuknf) [reply openziti.yml:22]: There are three containers in this compose file. One runs OpenZiti (this one). The other two will:
* chmod the named volume for writing
* initialize OpenZiti, after it's running

This is the "run OpenZiti" section and cannot be moved to after the "configure OpenZiti" section.

File: openziti-init-entrypoint.sh (hunk 5 above)
Comment (lenny-intel) [reply openziti-init-entrypoint.sh:65]: See https://github.com/edgexfoundry/edgex-compose/pull/440#discussion_r1538276301

File: openziti.yml (hunk 7 above)
Comment (lenny-intel) [reply openziti.yml:22]: Yep, figured that out... ;-)

File: openziti-init-entrypoint.sh (hunk 3 above)
Comment (dovholuknf) [reply openziti-init-entrypoint.sh:62]: is it expected for core services to call other core services at all?

Comment (lenny-intel) [reply openziti-init-entrypoint.sh:62]: Yes, Core Command calls on Core MetaData.

Also just remembered that Support schedule calls Core Data, but no thru the Core Data Clients.
See configuration here: https://github.com/edgexfoundry/edgex-go/blob/main/cmd/support-scheduler/res/configuration.yaml#L16-L20

The schedule can all on any endpoint of any service, not just EdgeX services. The configuration linked above just happens to be calling on Core Data. Do we simply need to env  override to host and port for this to work?

Comment (dovholuknf) [reply openziti-init-entrypoint.sh]: So the only core service that needs to communicate to another core service is core-command -> core-metadata and support-scheduler -> core-data but through the underlay? If that behavior is desired (using the underlay) technically the healthcheck proxy could be used for that, but it feels wrong to do that. It'd be better to adapt `scrub-aged-events` to use OpenZiti in some way.

At this time, I opted for core services to be authorized to dial other core services but that's easy to fix. The UI for example needs to dial all the core services, the two support services, two device services and an app service...

Have a look at the new openziti-init-entrypoint.sh specifically the new service-policy creation:
```

ziti edge create service-policy app-core-dial Dial --identity-roles "#application.id" --service-roles "#core.svc"
ziti edge create service-policy app-support-dial Dial --identity-roles "#application.id" --service-roles "#support.svc"
ziti edge create service-policy ds-core-dial Dial --identity-roles "#device.id" --service-roles "@edgex.core-metadata"
ziti edge create service-policy core-core-dial Dial --identity-roles "#core.id" --service-roles "#core.svc"
```

* app identities authorized to dial core services
* app identities authorized to dial support services
* device identities authorized to dial core metadata only
* core identities authorized to core support services

Comment (lenny-intel) [reply openziti-init-entrypoint.sh]: > At this time, I opted for core services to be authorized to dial other core services but that's easy to fix. The UI for example needs to dial all the core services, the two support services, two device services and an app service...

Actuall, I can't think of where the UI actually calls on the Device Services. If you found somewhere then it applies to all device service, not just the two in the current compose file.

File: docker-compose-zero-trust.yml [hunk 12]
```diff
@@ -0,0 +1,1136 @@
...
+        selinux: z
+        create_host_path: true
+  ui:
+    command:
+      - "./edgex-ui-server"
+      - "--configDir=res/docker"
+    container_name: edgex-ui-go
+    environment:
+      EDGEX_SECURITY_SECRET_STORE: "true"
+      SERVICE_HOST: 0.0.0.0
```
Comment (lenny-intel) [inline docker-compose-zero-trust.yml:961]: Why is this change needed?

File: docker-compose-zero-trust.yml [hunk 13]
```diff
@@ -0,0 +1,1136 @@
...
+      _SERVICE_PORT: 4000
```
Comment (lenny-intel) [inline docker-compose-zero-trust.yml:962]: Why the underscore? 4000 is already the Port value in the config file so no need for an override.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Comment by lenny-intel** [comment]
@dovholuknf , still need to add target to makefile to start/stop openzitit.

I also now think, as a checkpoint,  we should commit `docker-compose-zero-trust.yml` and add makefile target to start and stop it. 

Then we can work on the compose-builder option to dynamically build `docker-compose-zero-trust.yml` which is not going to be trivial... ;-)

File: openziti-init-entrypoint.sh [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+makeDeviceService 'device-usb-camera'
+makeDeviceService 'device-virtual'
+
+makeApplicationService 'app-external-mqtt-trigger'
+makeApplicationService 'app-http-export'
+makeApplicationService 'app-metrics-influxdb'
+makeApplicationService 'app-mqtt-export'
+makeApplicationService 'app-rfid-llrp-inventory'
+makeApplicationService 'app-rules-engine'
+makeApplicationService 'app-sample'
```
[the remark this answers is not in the document]
Comment (dovholuknf) [reply openziti-init-entrypoint.sh:76]: will address in next commit

File: openziti-init-entrypoint.sh [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+function makeApplicationService {
+  makeEdgeXService $1 "application"
+}
+
+sleep 1
+
+makeCoreService 'core-command'
+makeCoreService 'core-data'
+makeCoreService 'core-metadata'
+makeCoreService 'rules-engine'
```
[the remark this answers is not in the document]
Comment (dovholuknf) [reply openziti-init-entrypoint.sh:52]: do you want a blanket "all support services can talk to all core services" authorization or leave it ONLY for rules-engine -> core-command?

File: docker-compose-zero-trust.yml (hunk 13 above)
Comment (dovholuknf) [reply docker-compose-zero-trust.yml:961]: junk. will remove

File: docker-compose-zero-trust.yml (hunk 12 above)
Comment (dovholuknf) [reply docker-compose-zero-trust.yml:960]: the UI uses this service host to listen on the underlay. when it was set to 127.0.0.1 or frankly anything other than this - docker wouldn't pass the traffic through. I will look to see if i mistakenly use this in the UI listener. that's my guess.

Comment (dovholuknf) [reply docker-compose-zero-trust.yml]: reverted back to edgex-ui-go. thx!

File: openziti-init-entrypoint.sh (hunk 15 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply openziti-init-entrypoint.sh]: blanket

File: Makefile [hunk 16]
```diff
@@ -54,6 +54,30 @@ portainer:
...
+	${DOCKER_COMPOSE} -p openziti -f docker-compose-zero-trust.yml up -d
+
+zero-trust-down:
+	${DOCKER_COMPOSE} -p openziti -f docker-compose-zero-trust.yml down
+
+zero-trust-logs:
+	${DOCKER_COMPOSE} -p openziti -f docker-compose-zero-trust.yml logs -f
+
+zero-trust-clean:
+	${DOCKER_COMPOSE} -p openziti -f docker-compose-zero-trust.yml down -v
```
Comment (lenny-intel) [inline Makefile:79]: These should have `-p edgex`

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: Makefile (hunk 16 above)
Comment (dovholuknf) [reply Makefile]: all set. thx

**Review by lenny-intel — APPROVED** [review]
LGTM

**Comment by lenny-intel** [comment]
recheck

**Comment by lenny-intel** [comment]
@cherrycl , This PR is failing a single unrelated test. Actually this PR doesn't impact the Smoke Tests at all.
Can you check into it? THX!
https://jenkins.edgexfoundry.org/blue/organizations/jenkins/edgexfoundry%2Fedgex-taf-pipelines/detail/smoke-test/897/tests
